lib/sync.js

```diff
@@ -6,155 +6,96 @@ var _ = require('lodash'),
     i18nStrings = require('i18n-strings-files'),
     GS = require('google-spreadsheet'),
     stringFileLocator = require(path.resolve(__dirname, 'utils/stringFileLocator')),
-    updateStrings = require(path.resolve(__dirname, 'utils', 'updateStrings'));
+    updateStrings = require(path.resolve(__dirname, 'utils', 'updateStrings')),
+    Promise = require('bluebird'),
+
+    google = require(path.resolve(__dirname, 'providers', 'googleSpreadsheet')),
+    stringsUrl = require(path.resolve(__dirname, 'providers', 'stringsUrl'));
 
 var doubleEscape = function(string) {
     return string.replace(/[^\\]\\(n|r|t)/g, '\\\\$1');
 };
 
+var createProvider = function (config) {
+    switch (config.sync.provider) {
+        case "google":
+            return google;
+
+        case "url":
+            return stringsUrl;
+
+        default:
+            throw new Error('Unknown provider: ' + config.sync.provider);
+    }
+};
+
+var locateStringFiles = function (srcPath) {
+    return new Promise(function (resolve, reject) {
+        stringFileLocator(srcPath, resolve);
+    });
+};
+
 module.exports = function(srcPath, options) {
     var configFile = path.resolve(srcPath, options.config);
 
-
     fs.exists(configFile, function(exists) {
         if (exists) {
             readFile(path.resolve(process.cwd(), configFile), function(rawConfig) {
-                var config = JSON.parse(rawConfig), syncProvider, syncConfig,
-                    syncConfig, sheet, creds, keyColumn, valueColumns;
-
-                if (_.isObject(config['sync'])) {
-                    syncProvider = config['sync']['provider'];
-
-                    if (syncProvider === 'google') {
-                        syncConfig = config['sync']['config'];
-
-                        if (syncConfig.hasOwnProperty('file-id') && syncConfig.hasOwnProperty('creds')) {
-                            sheet = new GS(syncConfig['file-id']);
-                            creds = syncConfig['creds'];
-                        
-                            if (syncConfig.hasOwnProperty('columns')) {
-                                if (syncConfig['columns'].hasOwnProperty('key')) {
-                                    keyColumn = syncConfig['columns']['key'].replace(/\s+/g, '').toLowerCase();
-                                } else {
-                                    keyColumn = 'ioskey';
+                var config = JSON.parse(rawConfig);
+                var provider = createProvider(config);
+
+                provider.configure(config.sync.config)
+                    .then(function (config) {
+                        return locateStringFiles(srcPath)
+                            .then(function (stringFiles) {
+                                return Object.keys(stringFiles);
+                            })
+                            .then(function (languages) {
+                                return provider.getTranslations(config, languages);
```

Comment:
Yeah, though to be honest `configure` currently just verifies and processes the config, but doesn't mutate any state. Perhaps I should rename `.configure`, it kind of implies "I'm stateful and this is how you configure me", but that's not what's going on. How about `.processConfig`?